File: app/src/main/java/com/shakil/barivara/activities/auth/ForgotPasswordActivity.java
package com.shakil.barivara.activities.auth;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.shakil.barivara.R;
import com.shakil.barivara.databinding.ActivityForgotPasswordBinding;
import com.shakil.barivara.utils.Constants;
import com.shakil.barivara.utils.UX;
import com.shakil.barivara.utils.Validation;

import java.util.HashMap;
import java.util.Map;

import es.dmoral.toasty.Toasty;

public class ForgotPasswordActivity extends AppCompatActivity {
    private ActivityForgotPasswordBinding activityForgotPasswordBinding;
    private final Map<String, String[]> hashMap = new HashMap();
    private Validation validation;
    private FirebaseAuth firebaseAuth;
    private UX ux;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityForgotPasswordBinding = DataBindingUtil.setContentView(this, R.layout.activity_forgot_password);

        //region init objects and perform all UI interactions
        initUI();
        bindUiWithComponents();
        //endregion
    }

    //region init objects
    private void initUI() {
        ux = new UX(this);
        validation = new Validation(this, hashMap);
        firebaseAuth = FirebaseAuth.getInstance();
    }
    //endregion

    //region perform all UI interactions
    private void bindUiWithComponents() {
        //region validation
        validation.setEditTextIsNotEmpty(new String[]{"email"},
                new String[]{getString(R.string.email_validation)});
        //endregion

        //region reset email send listener job
        activityForgotPasswordBinding.changePassword.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (validation.isValid()){
                    ux.getLoadingView();
                    firebaseAuth.sendPasswordResetEmail(activityForgotPasswordBinding.email.getText().toString())
                            .addOnCompleteListener(new OnCompleteListener<Void>() {
                                @Override
                                public void onComplete(@NonNull Task<Void> task) {
                                    if (task.isSuccessful()) {
                                        Log.i(Constants.TAG, "Reset pass mail sent.");
                                        Toasty.info(ForgotPasswordActivity.this, getString(R.string.email_sent), Toast.LENGTH_SHORT, true).show();
                                        activityForgotPasswordBinding.changePasswordLayout.setVisibility(View.GONE);
                                        activityForgotPasswordBinding.afterChangePasswordLayout.setVisibility(View.VISIBLE);
                                    } else {
                                        Log.i(Constants.TAG, "Reset pass error : "+task.isSuccessful());
                                        if (task.getException().getMessage().equals(getString(R.string.firebase_no_user_exception))){
                                            Toasty.error(ForgotPasswordActivity.this, getString(R.string.email_was_not_found_in_our_database), Toast.LENGTH_SHORT, true).show();
                                        }
                                        else{
                                            Toasty.error(ForgotPasswordActivity.this, getString(R.string.failed_to_sent_reset_email), Toast.LENGTH_SHORT, true).show();
                                        }
                                        activityForgotPasswordBinding.changePasswordLayout.setVisibility(View.VISIBLE);
                                        activityForgotPasswordBinding.afterChangePasswordLayout.setVisibility(View.GONE);
                                    }
                                    ux.removeLoadingView();
                                }
                            });
                }
            }
        });
        //endregion

        activityForgotPasswordBinding.login.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(ForgotPasswordActivity.this, LoginActivity.class));
            }
        });
    }
    //endregion

    //region activity components
    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }
    //endregion
}

File: app/src/main/java/com/shakil/barivara/presentation/auth/forgotpassword/ForgotPasswordActivity.kt
package com.shakil.barivara.presentation.auth.forgotpassword

import android.os.Bundle
import androidx.activity.viewModels
import com.google.firebase.auth.FirebaseAuth
import com.shakil.barivara.BaseActivity
import com.shakil.barivara.R
import com.shakil.barivara.databinding.ActivityForgotPasswordBinding
import com.shakil.barivara.presentation.auth.AuthViewModel
import com.shakil.barivara.utils.Constants
import com.shakil.barivara.utils.PrefManager
import com.shakil.barivara.utils.UX
import com.shakil.barivara.utils.Validation
import dagger.hilt.android.AndroidEntryPoint
import es.dmoral.toasty.Toasty

@AndroidEntryPoint
class ForgotPasswordActivity : BaseActivity<ActivityForgotPasswordBinding>() {
    private lateinit var activityForgotPasswordBinding: ActivityForgotPasswordBinding
    private val hashMap: Map<String?, Array<String>?> = HashMap()
    private var validation = Validation(this, hashMap)
    private var firebaseAuth: FirebaseAuth? = null
    private lateinit var ux: UX
    private lateinit var prefManager: PrefManager
    private val viewModel by viewModels<AuthViewModel>()

    override val layoutResourceId: Int
        get() = R.layout.activity_forgot_password

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initUI()
        bindUiWithComponents()
        initListeners()
        initObservers()
    }

    override fun setVariables(dataBinding: ActivityForgotPasswordBinding) {
        activityForgotPasswordBinding = dataBinding
    }

    private fun initUI() {
        ux = UX(this)
        prefManager = PrefManager(this)
        firebaseAuth = FirebaseAuth.getInstance()
    }

    private fun bindUiWithComponents() {
        validation.setEditTextIsNotEmpty(
            arrayOf("oldPassword", "newPassword", "reEnterNewPassword"),
            arrayOf(
                getString(R.string.old_password_validation),
                getString(R.string.new_password_validation),
                getString(R.string.re_enter_password_validation)
            )
        )
    }

    private fun initListeners() {
        activityForgotPasswordBinding.changePassword.setOnClickListener {
            if (validation.isValid) {
                viewModel.changePassword(
                    prefManager.getString(Constants.mAccessToken),
                    activityForgotPasswordBinding.oldPassword.text.toString(),
                    activityForgotPasswordBinding.newPassword.text.toString(),
                    activityForgotPasswordBinding.reEnterNewPassword.text.toString()
                )
            }
        }
    }

    private fun initObservers() {
        viewModel.isLoading.observe(this) { isLoading ->
            if (isLoading) {
                ux.getLoadingView()
            } else {
                ux.removeLoadingView()
            }
        }

        viewModel.getChangePasswordResponse().observe(this) {
            if (it.statusCode == 200) {
                Toasty.success(this, getString(R.string.password_changed_successfully)).show()
            }
        }

        viewModel.getChangePasswordErrorResponse().observe(this) {
            Toasty.error(this, it.message).show()
        }
    }
}